File: 2015/day22.ts
interface Player {
    hp: number;
    damage: number;
    armor: number;
    mana: number;
    manaUsed: number;
    activeSpells: Record<string, number>;
}
interface Boss {
    hp: number;
    damage: number;
}
const spells = [
    {
        id: 'missile',
        cost: 53,
        cast: (player: Player, boss: Boss) => {
            player.mana -= 53;
            player.manaUsed += 53;
            boss.hp -= 4;
        },
        effect: () => {}
    },
    {
        id: 'drain',
        cost: 73,
        cast: (player: Player, boss: Boss) => {
            player.mana -= 73;
            player.manaUsed += 73;
            player.hp += 2;
            boss.hp -= 2;
        },
        effect: () => {}
    },
    {
        id: 'shield',
        cost: 113,
        cast: (player: Player) => {
            if (player.activeSpells.shield === undefined) {
                player.mana -= 113;
                player.manaUsed += 113;
                player.armor += 7;
                player.activeSpells.shield = 6;
                return;
            }
        },
        effect: (player: Player) => {
            player.activeSpells.shield -= 1;
            if (player.activeSpells.shield === 0) {
                player.armor -= 7;
                delete player.activeSpells.shield;
            }
        }
    },
    {
        id: 'poison',
        cost: 173,
        cast: (player: Player) => {
            if (player.activeSpells.poison === undefined) {
                player.mana -= 173;
                player.manaUsed += 173;
                player.activeSpells.poison = 6;
                return;
            }
        },
        effect: (player: Player, boss: Boss) => {
            boss.hp -= 3;
            player.activeSpells.poison -= 1;
            if (player.activeSpells.poison === 0) {
                delete player.activeSpells.poison;
            }
        }
    },
    {
        id: 'recharge',
        cost: 229,
        cast: (player: Player) => {
            if (player.activeSpells.recharge === undefined) {
                player.mana -= 229;
                player.manaUsed += 229;
                player.activeSpells.recharge = 5;
                return;
            }
        },
        effect: (player: Player) => {
            player.mana += 101;
            player.activeSpells.recharge -= 1;
            if (player.activeSpells.recharge === 0) {
                delete player.activeSpells.recharge;
            }
        }
    }
];

const boss = { hp: 55, damage: 8, armor: 0 };
const player = {
    hp: 50,
    damage: 0,
    armor: 0,
    mana: 500,
    manaUsed: 0,
    activeSpells: {}
};

function applyActiveSpells(player: Player, boss: Boss) {
    let keys = Object.keys(player.activeSpells);
    for (let i = 0; i < keys.length; i++) {
        spells.find((_s) => _s.id == keys[i])!.effect(player, boss);
        if (player.hp <= 0 || boss.hp <= 0) return;
    }
}

function play(
    _player: Player,
    _boss: Boss,
    spellId: string,
    hard: boolean
): [Player, Boss] {
    let player = {
        ..._player,
        activeSpells: { ..._player.activeSpells }
    };
    let boss = { ..._boss };
    let spell = spells.find((s) => s.id === spellId)!;

    //player turn
    if (hard) player.hp -= 1;
    if (player.hp <= 0) return [player, boss];
    applyActiveSpells(player, boss);
    spell.cast(player, boss);
    if (boss.hp <= 0) return [player, boss];

    //boss turn
    applyActiveSpells(player, boss);
    if (boss.hp <= 0) return [player, boss];
    player.hp -= Math.max(1, boss.damage - player.armor);
    return [player, boss];
}

function findLowestMana(player: Player, boss: Boss, hard: boolean = false) {
    let minManaUsed = Infinity;
    let queue = [{ boss, player }];
    while (queue.length) {
        let p = queue.pop()!;
        if (p.player.hp <= 0) continue;
        if (p.boss.hp <= 0) {
            if (p.player.manaUsed < minManaUsed) {
                minManaUsed = p.player.manaUsed;
            }
            continue;
        }
        if (p.player.manaUsed >= minManaUsed) continue;
        spells.forEach((s) => {
            if ((p.player.activeSpells[s.id] || 0) > 1) return;
            if (p.player.mana < s.cost) return;
            let [player, boss] = play(p.player, p.boss, s.id, hard);
            if (player.hp <= 0) return;
            queue.push({ player, boss });
        });
    }
    return minManaUsed;
}

console.log('Part 1', findLowestMana(player, boss));
console.log('Part 2', findLowestMana(player, boss, true));

File: 2015/day22.py
class Character():
    def __init__(self, hp, armor, dmg, mana):
        self.hp = hp
        self.armor = armor
        self.dmg = dmg
        self.mana = mana

class Insta():
    def __init__(self, dmg, heal):
        self.dmg = dmg
        self.heal = heal

    def perform(self, player, boss):
        player.hp += self.heal
        boss.hp -= self.dmg


class Effect():
    def __init__(self, id_, turns, armor_boost, dmg, mana_up):
        self.id_ = id_
        self.turns = turns
        self.armor_boost = armor_boost
        self.dmg = dmg
        self.mana_up = mana_up
        self.total_turns = turns

    def reset(self):
        return Effect(self.id_, self.total_turns, self.armor_boost, self.dmg, self.mana_up)

    def __hash__(self):
        return self.id_

    def __eq__(self, other):
        return self.id_ == other.id_

    def use(self, player, boss):
        if self.turns == self.total_turns:
            player.armor += self.armor_boost
        boss.hp -= self.dmg
        player.mana += self.mana_up
        self.turns -= 1
        if self.turns == 0:
            player.armor -= self.armor_boost

EMPTY_EFFECT = Effect(-1, 0, 0, 0, 0)

class Spell():
    def __init__(self, name, mana, effect, insta):
        self.name = name
        self.mana = mana
        self.effect = effect
        self.insta = insta

    def cast_spell(self, player, boss, effects):
        player.mana -= self.mana
        if player.mana < 0 or self.effect in effects:
            return False
        if self.insta is not None:
            self.insta.perform(player, boss)
        if self.effect != EMPTY_EFFECT:
            effects.add(self.effect)
        return True

def total_mana(spells):
    return sum([sp.mana for sp in spells])

def use_effects(effects, player, boss):
    for effect in effects:
        effect.use(player, boss)
    return set([e for e in effects if e.turns > 0])


def simulate_battle(spell_order):
    spell_order = [Spell(s.name, s.mana, s.effect.reset(), s.insta) for s in spell_order]
    boss = Character(55, 0, 8, 0)
    player = Character(50, 0, 0, 500)
    effects = set()
    sp = 0
    for sp in range(len(spell_order)):
        player.hp -= 1
        if player.hp <= 0:
            return -3
        effects = use_effects(effects, player, boss)
        if boss.hp <= 0:
            return total_mana(spell_order[:sp])
        if not spell_order[sp].cast_spell(player, boss, effects):
            return -2
        effects = use_effects(effects, player, boss)
        if boss.hp <= 0:
            return total_mana(spell_order[:sp+1])
        player.hp -= max(1, boss.dmg - player.armor)
        if player.hp <= 0:
            return -3
    return -1

spells = [
    Spell('Magic Missile', 53, EMPTY_EFFECT, Insta(4, 0)),
    Spell('Drain', 73, EMPTY_EFFECT, Insta(2, 2)),
    Spell('Shield', 113, Effect(1, 6, 7, 0, 0), None),
    Spell('Poison', 173, Effect(2, 6, 0, 3, 0), None),
    Spell('Recharge', 229, Effect(3, 5, 0, 0, 101), None)
]

sp_perm = [[]]
while True:
    n_sp = []
    for comb in sp_perm:
        for s in spells:
            n_sp.append(comb + [s])
    sp_perm = n_sp
    mn = -1
    next_sp_perm = []
    for sp in sp_perm:
        v = simulate_battle(sp)
        if v == -1:
           next_sp_perm.append(sp)
           continue
        if v == -2 or v == -3:
           continue
        if mn == -1 or v < mn:
            mn = v
    if mn != -1:
        break
    sp_perm = next_sp_perm
print(mn)
